### PR TITLE
Add introduction to CDEvents

### PR DESCRIPTION
**Pull request by m-linner-ericsson**
I wanted to create a introduction to CDEvents explaining to a larger audience why we created the project and what it provides. 

I have taken inspiration from previous PR material and other markdown documents, but also added a few idea myself.

If you feel the text should live in another document please feel free suggest that.

Signed-off-by: Mattias Linnér mattias.linner@ericsson.com

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,6 +6,88 @@ interoperability in the complete software production ecosystem.
...
+The larger and more complex a CI/CD system becomes, challenges increase in
+knowing how the tools communicate and what they do.
+
+### What we provide
+
+The CDEvents protocol defines a vocabulary of events enabling tools to
+communicate in a interoperable way.
+
+We extend other efforts such as CloudEvents by introducing purpose and semantics
+meaning of an event.
```
Comment (e-backmark-ericsson) [inline README.md:31]:
```suggestion
to the event.
```

File: README.md [hunk 2]
```diff
@@ -6,6 +6,88 @@ interoperability in the complete software production ecosystem.
...
+
+We thus build on top of the work from other players as the following stack
+pictures exemplifies.
+
+![stack](./images/stack.png)
+
+By providing a interoperable way of tool to communicate we also provide means to
```
Comment (e-backmark-ericsson) [inline README.md:38]:
```suggestion
By providing an interoperable way of tools to communicate we also provide means to
```

File: README.md [hunk 3]
```diff
@@ -6,6 +6,88 @@ interoperability in the complete software production ecosystem.
...
+give an overview picture increasing observability, but also provides measuring
```
Comment (e-backmark-ericsson) [inline README.md:39]:
```suggestion
give an overview picture increasing observability, but also to give measuring
```

File: README.md [hunk 4]
```diff
@@ -6,6 +6,88 @@ interoperability in the complete software production ecosystem.
...
+points for metrics.
+
+#### How does it enable tools to communicate in a interoperable way?
+
+By creating a language, we define how tools and services communicate with each
+other about occurrences in a CI/CD system. As this language does not tie to a
+specific tool it serves a neutral ground for communication.
+
+Using this language we defining a set of events that has purpose and semantic
```
Comment (e-backmark-ericsson) [inline README.md:48]:
```suggestion
Using this language we define a set of events with purpose and semantic
```

File: README.md [hunk 5]
```diff
@@ -6,6 +6,88 @@ interoperability in the complete software production ecosystem.
...
+We believe that using events will lead to a more decoupled systems with services
+and tools developed and deployed independently. This makes us agnostic of the
+underlying infrastructure
+
+#### Why not point-to-point communication?
+
+We believe that using integrations based on point-to-point communication will
+create a system that will:
+
+* Not scale - when trying to add new consumers or senders each tools have to make
```
Comment (e-backmark-ericsson) [inline README.md:77]:
```suggestion
* Not scale - when trying to add new consumers or producers each tool have to make
```

**Review by e-backmark-ericsson — CHANGES_REQUESTED** [review]
I embrace this initiative, and I believe it captures the background well enough as it is now written.

Maybe this page deserves a ToC now as its content has grown quite a lot?

**Comment by afrittoli** [comment]
> I wanted to create a introduction to CDEvents explaining to a larger audience why we created the project and what it provides.
> 
> I have taken inspiration from previous PR material and other markdown documents, but also added a few idea myself.
> 
> If you feel the text should live in another document please feel free suggest that.
> 
> Signed-off-by: Mattias Linnér [mattias.linner@ericsson.com](mailto:mattias.linner@ericsson.com)

Thanks @m-linner-ericsson !
I think it's nice to provide a little bit more context in the main README. But I would prefer to keep the READM not too busy, and maybe move some of this content to the CDEvents primer?

File: images/stack.drawio [hunk 6]
```diff
@@ -0,0 +1 @@
+<mxfile host="Electron" modified="2022-03-09T14:09:43.285Z" agent="5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) draw.io/16.5.1 Chrome/96.0.4664.110 Electron/16.0.7 Safari/537.36" etag="WJztc4Q_z2C4iVk_88a6" version="16.5.1" type="device"><diagram id="rZAp1BOKmzg2OGIajMNH" name="Page-1">5ZbdbqMwEIWfhsuVAEM2uUwJTaRVpLap1N1LC0/AG2MjY0rSp69dTAhxK3Wl7Y9aiQv7zNhmvmNsPJSU+6XEVbEWBJgX+mTvoYUXhkEUhp55fHLolJkfd0IuKbFJg7ChD2BF36oNJVCPEpUQTNFqLGaCc8jUSMNSinacthVsvGqFc3CETYaZq95RoopOncb+oK+A5kW/cuDbSIn7ZCvUBSaiPZFQ6qFECqG6VrlPgBl4PZdu3OUL0eOLSeDqNQOK9E+6WM5n02v/7m9THbLfy/aHneUes8YWvLq9vfLCRIu/8HaHu+Z8fX1lq1CHHo0UDSdgZg88dNEWVMGmwpmJtnozaK1QJbNhuw5IBfsXCwiOWPR+AlGCkgedYgegnqTdSnFk++1gTNDnFCemTKyG7V7Ij1MPuHTDEvsHeqFDL2GiIem9rqj+dLyi6UfzQi6vxWeFFX40rNiBdSMaRXnusNJVqzGQWkmxg0QwIbXCBdeZF1vK2JmEGc257jLYmhkMQarPvrmVS0qIWeRZ/oND/v+xIJqcfd9x7FgwecYB9FYOTBwH1lDX5sIwF4nUJ/x3sSKauVYcb7p38eKn48UGSsw1I/fs+KomoLczQXeHP5Gn2Mn/HEofAQ==</diagram></mxfile>
```
Comment (afrittoli) [inline images/stack.drawio:1]: Thanks for including the source!

File: README.md [hunk 7]
```diff
@@ -6,6 +6,88 @@ interoperability in the complete software production ecosystem.
...
+We thus build on top of the work from other players as the following stack
+pictures exemplifies.
+
+![stack](./images/stack.png)
+
+By providing a interoperable way of tool to communicate we also provide means to
+give an overview picture increasing observability, but also provides measuring
+points for metrics.
+
+#### How does it enable tools to communicate in a interoperable way?
```
[the remark this answers is not in the document]
Comment (afrittoli) [reply README.md:42]: Perhaps from this point on we could add this material to the `primer.md` document?

File: README.md [hunk 8]
```diff
@@ -6,6 +6,88 @@ interoperability in the complete software production ecosystem.
...
+The CDEvents protocol defines a vocabulary of events enabling tools to
+communicate in a interoperable way.
+
+We extend other efforts such as CloudEvents by introducing purpose and semantics
+meaning of an event.
+
+We thus build on top of the work from other players as the following stack
+pictures exemplifies.
+
+![stack](./images/stack.png)
```
Comment (afrittoli) [inline README.md:36]: It looks like something went wrong with the rendering of this diagram. The font looks rather jagged...

File: spec.md [hunk 9]
```diff
@@ -14,7 +14,8 @@ of the CDEvents project. You are very welcome to
 
 ## Abstract
 
-CDEvents is a common specification for Continuous Delivery events.
+CDEvents is a common specification for Continuous Delivery events. For more
```
Comment (afrittoli) [inline spec.md:17]: I would not add this to the abstract.
We could link the README and primer docs in the overview below instead?

File: README.md [hunk 10]
```diff
@@ -6,6 +6,88 @@ interoperability in the complete software production ecosystem.
...
+
+We believe that using integrations based on point-to-point communication will
+create a system that will:
+
+* Not scale - when trying to add new consumers or senders each tools have to make
+  an update
+* Create a coupled architecture - using point-to-point communication creates a
+  tightly intertwined architecture difficult to expand and monitor.
+
+#### Why not webhooks?
```
Comment (afrittoli) [inline README.md:82]: I'm not sure about this. I see webhooks as a way to deliver events, and I didn't know we recommended explicitly against them. CloudEvents include a spec for webhooks https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/http-webhook.md and they could be an interesting option in cases where authorization is required.

File: README.md [hunk 11]
```diff
@@ -6,6 +6,88 @@ interoperability in the complete software production ecosystem.
...
+### What we provide
+
+The CDEvents protocol defines a vocabulary of events enabling tools to
+communicate in a interoperable way.
+
+We extend other efforts such as CloudEvents by introducing purpose and semantics
+meaning of an event.
+
+We thus build on top of the work from other players as the following stack
+pictures exemplifies.
```
[the remark this answers is not in the document]
Comment (m-linner-ericsson) [reply README.md]: Updated in 1ad4def

File: images/stack.drawio (hunk 6 above)
Comment (m-linner-ericsson) [reply images/stack.drawio:1]: 😃

File: spec.md (hunk 9 above)
Comment (m-linner-ericsson) [reply spec.md]: Tried to make something in 1ad4def. Do you have any other good ideas?

File: README.md (hunk 10 above)
Comment (m-linner-ericsson) [reply README.md]: Felt this was a bit on the edge so I removed it in 1ad4def

File: README.md (hunk 8 above)
Comment (m-linner-ericsson) [reply README.md:33]: Hmm, that's not good. I however don't see the problem do you have some more info?

Comment (afrittoli) [reply README.md:33]: ok no worries, I don't want to block the PR on this - I see you included the source as well so I can also try to re-render it on my side

File: spec.md [hunk 12]
```diff
@@ -25,6 +25,9 @@ The specification is structured in two main parts:
 - The [*vocabulary*](vocabulary), which identifies *event types*, structures as
   *subjects* and *predicates*
 
+For an introduction see [README.md](README.md) and for  more background
```
Comment (afrittoli) [inline spec.md:28]: NIT: extra space "for  more"

**Review by afrittoli — APPROVED** [review]
Thank for all the updates.
Just a nit but looks good to me

File: spec.md (hunk 12 above)
Comment (m-linner-ericsson) [inline spec.md:28]:
```suggestion
For an introduction see [README.md](README.md) and for more background
```

Comment (m-linner-ericsson) [reply spec.md]: Fixed in 5bea30c

**Comment by afrittoli** [comment]
@e-backmark-ericsson ok to merge?

File: primer.md [hunk 13]
```diff
@@ -17,6 +17,7 @@ This non-normative document provides an overview of the CDEvents specification.
 <!-- toc -->
 - [History](#history)
 - [Relations to CloudEvents](#relations-to-cloudevents)
+- [Design reflections](#design-reflections)
```
Comment (e-backmark-ericsson) [inline primer.md:20]: This ToC entry should be swapped with the previous line to match the actual content

File: README.md [hunk 14]
```diff
@@ -6,6 +6,36 @@ interoperability in the complete software production ecosystem.
...
+platforms that need to communicate with each other interoperability stands as a
+crucial thing. The maintainer of a CI/CD system needs to swap out tools in short
+time with little to no stops.
+
+The larger and more complex a CI/CD system becomes, challenges increase in
+knowing how the tools communicate and what they do.
+
+### What we provide
+
+The CDEvents protocol defines a vocabulary of events enabling tools to
```
Comment (e-backmark-ericsson) [inline README.md:27]: As this is the first time "events" is mentioned in this document (which is probably the first document people read on CDEvents), I believe we should either link to a description of what an event is here, or add a sentence/paragraph about it inline

**Review by e-backmark-ericsson — CHANGES_REQUESTED** [review]
Just a very small change needed

File: primer.md (hunk 13 above)
Comment (afrittoli) [reply primer.md:20]: You can run `./tools/update-toc.sh` to keep the ToCs updated - I should probably setup a CI job for that :)

Comment (m-linner-ericsson) [reply primer.md:19]: Should be fixed in 610d408

**Review by e-backmark-ericsson — APPROVED** [review]

**Pull request merged**